src/Parsers/ASTColumnsTransformers.cpp

```diff
@@ -61,28 +86,49 @@ void ASTColumnsExceptTransformer::formatImpl(const FormatSettings & settings, Fo
         (*it)->formatImpl(settings, state, frame);
     }
 
-    settings.ostr << ")";
+    if (children.size() > 1)
+        settings.ostr << ")";
 }
 
 void ASTColumnsExceptTransformer::transform(ASTs & nodes) const
 {
+    ASTs expected_columns(children);
+
     nodes.erase(
         std::remove_if(
             nodes.begin(),
             nodes.end(),
-            [this](const ASTPtr & node_child)
+            [&](const ASTPtr & node_child)
             {
                 if (const auto * id = node_child->as<ASTIdentifier>())
                 {
-                    for (const auto & except_child : children)
+                    for (int i = expected_columns.size() - 1; i >= 0; --i)
                     {
-                        if (except_child->as<const ASTIdentifier &>().name() == id->shortName())
+                        if (expected_columns[i]->as<const ASTIdentifier &>().name() == id->shortName())
+                        {
+                            expected_columns.erase(expected_columns.begin() + i);
```

Comment:
So you're modifying some other data structure from the predicate of `remove_if`? Looks weird but probably OK...

Reply:
I want to get the columns that except specified but not in table's real columns, and try to report the error if it is a strict except query, 
If this kind of coding is not ok, I will try to  use another way.

Reply:
It's hard for readers to reason about what's happening. It would be better to use a simple for loop.